## Baysian Topic Segmentation . 

This code is based on analysis on the basis Baysian Unsupervised Topic segmentation 2008 by Jacob Eisentien and Regina Barzilay, Computer Science and Artificial Intelligence Laboratory , Massachusetts Institute of Technology. 
The code and paper is refered from the following [ link](http://groups.csail.mit.edu/rbg/code/bayesseg/bayesseg.tar.gz)

# Downloading and evaluating the base code

In [0]:
!wget http://groups.csail.mit.edu/rbg/code/bayesseg/bayesseg.tar.gz

--2019-09-03 05:37:21--  http://groups.csail.mit.edu/rbg/code/bayesseg/bayesseg.tar.gz
Resolving groups.csail.mit.edu (groups.csail.mit.edu)... 128.30.2.44
Connecting to groups.csail.mit.edu (groups.csail.mit.edu)|128.30.2.44|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4687672 (4.5M) [application/x-gzip]
Saving to: ‘bayesseg.tar.gz’

bayesseg.tar.gz     100%[===================>]   4.47M  23.1MB/s    in 0.2s    

2019-09-03 05:37:21 (23.1 MB/s) - ‘bayesseg.tar.gz’ saved [4687672/4687672]



In [0]:
!tar zxvf bayesseg.tar.gz

bayesseg/baselines/
bayesseg/baselines/textseg-1.211/
bayesseg/baselines/textseg-1.211/data/
bayesseg/baselines/textseg-1.211/data/comp/
bayesseg/baselines/textseg-1.211/data/org/
bayesseg/baselines/textseg-1.211/data/t/
bayesseg/baselines/textseg-1.211/doc/
bayesseg/baselines/textseg-1.211/doc/eng/
bayesseg/classes/
bayesseg/classes/edu/
bayesseg/classes/edu/mit/
bayesseg/classes/edu/mit/multimodal/
bayesseg/classes/edu/mit/multimodal/motifs/
bayesseg/classes/edu/mit/nlp/
bayesseg/classes/edu/mit/nlp/segmenter/
bayesseg/classes/edu/mit/nlp/segmenter/dp/
bayesseg/classes/edu/mit/nlp/segmenter/mcmc/
bayesseg/classes/edu/mit/nlp/segmenter/wrappers/
bayesseg/classes/edu/mit/util/
bayesseg/classes/edu/mit/util/ling/
bayesseg/classes/edu/mit/util/stats/
bayesseg/classes/edu/mit/util/weka/
bayesseg/classes/riso/
bayesseg/classes/riso/numerical/
bayesseg/config/
bayesseg/data/
bayesseg/data/books/
bayesseg/data/books/clinical/
bayesseg/data/icsi/
bayesseg/data/icsi/scripts/
bayesseg/data/icsi

In [0]:
%cd bayesseg

/content/bayesseg


In [0]:
!chmod 700 eval

In [0]:
!./eval config/dp.config

{use-extra-features=false, careful-debug=false, segmentation-dispersion-range=.1,100,20,M, dirichlet-prior=.2, window-size-range=10,25,16,A, ;;;;;;;;;;=MCMC PARAMS, use-fixed-blocks=false, use-word-stems=true, update-params-period=1000, mcem-cuephrases=false, lambda-b=1, max-move=50, use-choi-style-boundaries=true, segmentation-dispersion=1, remove-stop-words=true, segmenter=edu.mit.nlp.segmenter.dp.BayesWrapper, window-size=10, cooling-duration=.25, ;;=em parameters already plugged in..., num-mcmc-moves=100000, output-period=100, cuephrase-file=config/CUEPHRASES.hl, max-burnin-temp=5, phi-b-0=.01, lambda-o=0, update-lm-period=100, use-duration=false, dirichlet-prior-range=.0001,.1,20,M, em-param-search=false, num-segs-known=true, stop-words=config/STOPWORD.list, phi-o-0=.005, burnin-duration=.25}
data/books/clinical/161.ref
data/books/clinical/141.ref
data/books/clinical/121.ref
data/books/clinical/101.ref
data/books/clinical/051.ref
data/books/clinical/061.ref
data/books/clinical/001

# New Dataset Creation

## Clonning for repo based on original data 

In [0]:
!git clone https://snj-adhikari:culblank1minded@github.com/snj-adhikari/bayesian-text-segmentation

Cloning into 'bayesian-text-segmentation'...
remote: Enumerating objects: 56, done.
remote: Counting objects: 100% (56/56), done.
remote: Compressing objects: 100% (36/36), done.
remote: Total 1092 (delta 27), reused 40 (delta 16), pack-reused 1036
Receiving objects: 100% (1092/1092), 8.19 MiB | 25.96 MiB/s, done.
Resolving deltas: 100% (372/372), done.


In [0]:
%cd bayesian-text-segmentation

/content/bayesian-text-segmentation


In [0]:
!chmod 700 eval


## Gather data using wikipedia


The document came fro wikipedia . 
Got all the links from the following wikipedia article
https://en.wikipedia.org/wiki/Music

- Took first 20 links from this article and got seperate article for that link to make our two document. 

- Crated 20 document named *.ref ranging from 0 - 19 , one document for one article from the link on wiki , 
those document where segmented using 20 to 30 lines segmentation using choi notation.

- Random sample *.ref from  0 - 10 , to get 9 articles document. Combined those 9 articles to make single file 1.txt 
- Again random sample *.ref from 10-19 to get 9 articles document . Combine those 9 articles to make another single file 2.txt
 
 - Finally , composite document was made by combining these two documents , splitting them up using "choi notation" iteslf,
 and interleaving the segments.

### Installing the wikipedia

In [0]:
!pip install wikipedia

  Created wheel for wikipedia: filename=wikipedia-1.4.0-cp36-none-any.whl size=11686 sha256=84d1d028358e63f96d89ad0b5771a22210755a581a0b1d81728377c6b1d85672
  Stored in directory: /root/.cache/pip/wheels/87/2a/18/4e471fd96d12114d16fe4a446d00c3b38fb9efcb744bd31f4a
Successfully built wikipedia


### Downloading NLTK with punkt for tokenize the data

In [0]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

### Creating Refrence document using individual wiki page

In [0]:
!rm -r data/music
!mkdir data/music


Basically , here we took music wikipedia page to extract all the links based on the music wiki page. From their I took first 20 links and extracted content from those link itself. Then,I used to tokenize package to extract sentences from the wikipedia page. With the help of that I created 20 pages .ref files. Then randomly selected 20 to 30 lines interval and segmented using choi notation.  As a result , the data/music file has now 20 files , named from 0.ref to 20.ref

note : some of the segment might have lesser than 20 sentences

In [0]:
import wikipedia as wiki
import random
import nltk.data

tokenizer = nltk.data.load('nltk:tokenizers/punkt/english.pickle')

music = wiki.page("music").links


step = 0 
for i in music[:20]:
    base_data = ""
    try:
      p = wiki.page(i)
    except wiki.exceptions.DisambiguationError as e:
      s = random.choice(e.options)
#       search = wiki.search(s)
#       s = random.choice(search)
#       print("the ambigious is " , s)
      p = wiki.page(s)
    except wiki.exceptions.PageError as e:
#       print("error on getting page")
      search= wiki.search(music[i])
      s = random.choice(search)
      p = wiki.page(s)
      
    new_data = p.content
    
    analyzed_data = new_data.replace("=" , " --")
#     print("the aanlayzed data is" , len(analyzed_data.split()))
    tokenized_data  = tokenizer.tokenize(analyzed_data)
    suffle_sen = random.randint(20,30)
#     print("the suffle sen is" , suffle_sen)
    step_token= len(tokenized_data)//suffle_sen
    
#     print("the step is" , step_token)
    for count in range(step_token) : 
      segmented_data = ''.join(tokenized_data[count *suffle_sen : (count+1) * suffle_sen]) 
      base_data = base_data + segmented_data +  "\n==========\n"
    
    if step_token <= 0 :
      segmented_data = ''.join(tokenized_data)
      base_data = base_data + segmented_data + "\n==========\n"
     
    words = len(base_data.split())
  
    text_file = open('data/music/'+str(step)+'.ref', 'w' )
    step  = step+ 1
    text_file.write(base_data)
    text_file.close()

    print("paragraph words written in file is " , words)
      

paragraph words written in file is  1944
paragraph words written in file is  2620
paragraph words written in file is  2372
paragraph words written in file is  895
paragraph words written in file is  1388
paragraph words written in file is  1816
paragraph words written in file is  3310
paragraph words written in file is  627
paragraph words written in file is  4081
paragraph words written in file is  14182
paragraph words written in file is  435
paragraph words written in file is  853


/usr/local/lib/python3.6/dist-packages/wikipedia/wikipedia.py:389: UserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("lxml"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 389 of the file /usr/local/lib/python3.6/dist-packages/wikipedia/wikipedia.py. To get rid of this warning, pass the additional argument 'features="lxml"' to the BeautifulSoup constructor.

  lis = BeautifulSoup(html).find_all('li')


paragraph words written in file is  572
paragraph words written in file is  2979
paragraph words written in file is  388
paragraph words written in file is  998
paragraph words written in file is  3588
paragraph words written in file is  2498
paragraph words written in file is  4856
paragraph words written in file is  10377


### Checking the segmentation of reference file.

In [0]:
!chmod 700 segment
!cat "data/music/1.ref" | ./segment config/dp.config

[1, 20, 27, 34]


## Creating new dataset document

### First creating two different text document 

Created a composite directory , made 2 composite document with random sample of  9 files between range(10) and range(10,20)Basically first range give files from 0.ref to 9.ref and second range gives file from 10.ref  to 19.ref 

In [0]:
!rm -r data/composite
!mkdir data/composite

In [0]:
import random 
total_no_first = range(10)
total_no_second = range(10,20)


sample_1 = random.sample(total_no_first, k=9)
sample_2 = random.sample(total_no_second, k=9)

data = ""
for i in sample_1:
  f = open("data/music/"+str(i)+".ref", "r")
  data = data + f.read()
  f.close()
  
  
print("the data length is" , len(data.split()))
new_file = open("data/composite/1.txt" , 'w')

new_file.write(data)
  
new_file.close()
  
  
data = ""
for i in sample_2 :
  f = open("data/music/"+str(i)+".ref", "r")
  data = data + f.read()
  f.close()
  
  
print("the data length is" , len(data.split()))
  
  
new_file = open("data/composite/2.txt" , 'w')

new_file.write(data)

new_file.close()
  

  

the data length is 32340
the data length is 22688


note : Got two document 1.txt and 2.txt basically around 24000 words to 35000 words each.

### Making the composite document , merging two document

Now taking two composite file creating making combined  document  using alternative segment from each file to combined document.

In [0]:
!rm -r data/combined
!mkdir data/combined

Create your artificial dataset by taking two documents i.e. 1.txt and 2.txt combined with different topics, splitting them up, and interleaving the segments.  

doc1_seg1 doc2_seg1 doc1_seg2 doc2_seg2 ... doc1_segN doc2_s

where 
doc1 = 1.txt  
seg = segment seperated by choi notation in previous data formation 

doc 2 = 2.txt ,  

In [0]:
array_split = []
f = open("data/composite/1.txt", "r")
data_1 =  f.read()
array_split_1 = data_1.split('==========')
f.close()


f2 = open("data/composite/2.txt", "r")
data_2 =  f2.read()
array_split_2 = data_2.split('==========')
f2.close()
  
  
print("the array split data is" , len(array_split_1))
print("the array split second data is" , len(array_split_2))
new_data = "==========\n"
if(len(array_split_1)>len(array_split_2)):
  for first in range(0, len(array_split_2)-1):
    new_data = new_data + array_split_1[first] + "==========\n" + array_split_2[first] + "==========\n"
else : 
  for second in range(0, len(array_split_1)-1):
    new_data = new_data + array_split_1[second] + "==========\n" + array_split_2[second] + "==========\n"


new_comp_file = open("data/combined/composite.txt" , 'w')

new_comp_file.write(new_data)
  
new_file.close()

the array split data is 54
the array split second data is 42


In [0]:
!git pull

remote: Enumerating objects: 5, done.
remote: Counting objects: 100% (5/5), done.
remote: Compressing objects: 100% (1/1), done.
remote: Total 3 (delta 2), reused 3 (delta 2), pack-reused 0
Unpacking objects: 100% (3/3), done.
From https://github.com/snj-adhikari/bayesian-text-segmentation
   cd4aa5f..764cfb6  master     -> origin/master
Updating cd4aa5f..764cfb6
Fast-forward
 eval | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)


## Testing new dataset

In [0]:
!chmod 700 eval
!./eval config/dp.config

{use-extra-features=false, careful-debug=false, segmentation-dispersion-range=.1,100,20,M, dirichlet-prior=.2, window-size-range=10,25,16,A, ;;;;;;;;;;=MCMC PARAMS, use-fixed-blocks=false, use-word-stems=true, update-params-period=1000, mcem-cuephrases=false, lambda-b=1, max-move=50, use-choi-style-boundaries=true, segmentation-dispersion=1, remove-stop-words=true, segmenter=edu.mit.nlp.segmenter.dp.BayesWrapper, window-size=15, cooling-duration=.25, ;;=em parameters already plugged in..., num-mcmc-moves=100000, output-period=100, cuephrase-file=config/CUEPHRASES.hl, max-burnin-temp=5, phi-b-0=.01, lambda-o=0, update-lm-period=100, use-duration=false, dirichlet-prior-range=.0001,.1,20,M, em-param-search=false, num-segs-known=true, stop-words=config/STOPWORD.list, phi-o-0=.005, burnin-duration=.25}
data/combined/composite.txt
0.4184 0.4920
0.4184 0.4920




The result  is similar to the  baysian paper. 

The result is average , the idea result was 0 for perfect segmentation. 
But our result is arround 0.5 for windDiff and 0.44 for pk which we can consider ideal for random segmentation.
Since our segmentation is random it is giving WinDiff around 0.5  i.e 0.49 

while pk is close to 0.4184 . So we are some what successfull to make random segmentation rather than the ideal segmentation 
with our new dataset. 

Also , PK will be less than winDiff same as the paper. 

Hence , we were able to replicate the bayesian topic segmentation code with the new data successfully.

**To conclude , we could consider the result is sufficiently good enough for random segmentation.**

## Pushing the changes into git so that it can be use be used in amazon VM

In [0]:
!git config --global user.email "adhikari.snj@gmail.com"
!git config --global user.name "Sanjay Adhikari"

In [0]:
!git add *
!git commit -m "ongoing changes"
!git push

[master dce41a3] ongoing changes
 1 file changed, 40 insertions(+), 40 deletions(-)
Counting objects: 5, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (4/4), done.
Writing objects: 100% (5/5), 659 bytes | 659.00 KiB/s, done.
Total 5 (delta 3), reused 0 (delta 0)
remote: Resolving deltas: 100% (3/3), completed with 3 local objects.
To https://github.com/snj-adhikari/bayesian-text-segmentation
   ad863f3..dce41a3  master -> master
